perf/src/packet.rs

```diff
@@ -13,21 +16,599 @@ use {
+            Self::Pinned(batch) => batch.get(index).map(PacketRef::from),
+            Self::Bytes(batch) => batch.get(index).map(PacketRef::from),
+        }
+    }
+
+    pub fn get_mut(&mut self, index: usize) -> Option<PacketRefMut<'_>> {
+        match self {
+            Self::Pinned(batch) => batch.get_mut(index).map(PacketRefMut::from),
+            Self::Bytes(batch) => batch.get_mut(index).map(PacketRefMut::from),
+        }
+    }
+
+    pub fn iter(&self) -> PacketBatchIter<'_> {
+        match self {
+            Self::Pinned(batch) => PacketBatchIter::Pinned(batch.iter()),
+            Self::Bytes(batch) => PacketBatchIter::Bytes(batch.iter()),
+        }
+    }
+
+    pub fn iter_mut(&mut self) -> PacketBatchIterMut<'_> {
+        match self {
+            Self::Pinned(batch) => PacketBatchIterMut::Pinned(batch.iter_mut()),
+            Self::Bytes(batch) => PacketBatchIterMut::Bytes(batch.iter_mut()),
+        }
+    }
+
+    pub fn par_iter(&self) -> PacketBatchParIter {
+        match self {
+            Self::Pinned(batch) => {
+                PacketBatchParIter::Pinned(batch.par_iter().map(PacketRef::from))
+            }
+            Self::Bytes(batch) => PacketBatchParIter::Bytes(batch.par_iter().map(PacketRef::from)),
+        }
+    }
+
+    pub fn par_iter_mut(&mut self) -> PacketBatchParIterMut {
+        match self {
+            Self::Pinned(batch) => {
+                PacketBatchParIterMut::Pinned(batch.par_iter_mut().map(PacketRefMut::from))
+            }
+            Self::Bytes(batch) => {
+                PacketBatchParIterMut::Bytes(batch.par_iter_mut().map(PacketRefMut::from))
+            }
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        match self {
+            Self::Pinned(batch) => batch.len(),
+            Self::Bytes(batch) => batch.len(),
+        }
+    }
+}
+
+impl From<PinnedPacketBatch> for PacketBatch {
+    fn from(batch: PinnedPacketBatch) -> Self {
+        Self::Pinned(batch)
+    }
+}
+
+impl From<BytesPacketBatch> for PacketBatch {
+    fn from(batch: BytesPacketBatch) -> Self {
+        Self::Bytes(batch)
+    }
+}
+
+impl<'a> IntoIterator for &'a PacketBatch {
+    type Item = PacketRef<'a>;
+    type IntoIter = PacketBatchIter<'a>;
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
+impl<'a> IntoIterator for &'a mut PacketBatch {
+    type Item = PacketRefMut<'a>;
+    type IntoIter = PacketBatchIterMut<'a>;
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter_mut()
+    }
+}
+
+impl<'a> IntoParallelIterator for &'a PacketBatch {
+    type Iter = PacketBatchParIter<'a>;
+    type Item = PacketRef<'a>;
+    fn into_par_iter(self) -> Self::Iter {
+        self.par_iter()
+    }
+}
+
+impl<'a> IntoParallelIterator for &'a mut PacketBatch {
+    type Iter = PacketBatchParIterMut<'a>;
+    type Item = PacketRefMut<'a>;
+    fn into_par_iter(self) -> Self::Iter {
+        self.par_iter_mut()
+    }
+}
+
+#[derive(Debug, Eq)]
```

Comment:
I wonder if should derive other traits that `PacketBatch` derives

Reply:
`PacketRef` is a wrapper for a reference of a single packet, not a whole batch, so probably not.